ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Adversarial.hs

```diff
@@ -249,20 +252,44 @@ checkAdversarialChain recipe adv = do
                 Nothing     -> pure ()   -- there are no remaining honest active slots
                                          --
                                          -- TODO hpc shows this never executes
-    checkDensity :: Exn.Except (AdversarialViolation hon adv) ()
-    checkDensity =
-        when (C.Count s <= C.windowSize winA) $ do
-            let -- first stability window after the intersection in the honest
-                -- schema
-                hScgWin = C.UnsafeContains (C.Count $ C.getCount $ C.windowStart winA) (C.Count s)
-                -- first stability window after the intersection in the
-                -- alternative schema
-                aScgWin = C.UnsafeContains (C.Count 0) (C.Count s)
-                -- number of active slots in the first stability window after
-                -- the intersection in the honest schema
-                hCount = C.getCount $ BV.countActivesInV S.notInverted (C.sliceV hScgWin vH)
-                aCount = C.getCount $ BV.countActivesInV S.notInverted (C.sliceV aScgWin vA)
-            when (aCount >= hCount) $ Exn.throwError $ BadDensity hCount aCount
+
+    -- | Check that the density of the adversarial schema is less than the
+    -- density of the honest schema in the first stability window after the
+    -- intersection and in any subwindow that contains the first race to the
+    -- k+1st block.
+    --
+    -- See description of @ensureLowerDensityInWindows@
+    checkDensity = do
+        let
+          -- window of the honest schema after the intersection
+          winHAfterIntersection =
+              C.UnsafeContains
+                  (fromJust $ C.toWindow winH $ C.windowStart winA)
+                  (C.Count $ C.getCount (C.windowSize winA))
+          -- honest schema after the intersection
+          vHAfterIntersection = C.sliceV winHAfterIntersection vH
+          iterH :: RI.Race adv
+          iterH =
+              maybe (RI.initConservative vHAfterIntersection) id
```

Comment:
We definitely want to include the "conservative" windows, ie the ones that assume the honest schema after the end of time is full of 1s?

Reply:
Yeah, interpreting this as a postcondition of `uniformAdversarialChain` does suffice to justify it.

---

> Check that the density of the adversarial schema is less than the
> density of the honest schema in the first stability window after the
> intersection and in any prefix that contains the first race to the
> k+1st block.

The use of `initConservative` here means we're doing this check even if there aren't `k+1` _actual_ active slots after the intersection in the schema. (And, indirectly, it also means that `uniformAdversarialChain` is.)

I suppose that's fine, merely hard to think about (as considering slots past the end of the schemas is always confusing :/). Perhaps it's somewhat "wasteful" when `RI.init` returns `Nothing`, since I _think_ in that case this check simplifies down to requiring that the honest schema has at least as many active slots after the intersection as the adversarial schema does. But "wasteful" is lower priority than "clear", and using the same predicate regardless of the ctor returned by `RI.init` is probably clearer.

Maybe it's worth a comment, if you agree with my conjecture about the simpler interpretation?

Reply:
> in that case this check simplifies down to requiring that the honest schema has at least as many active slots after the intersection as the adversarial schema does.

I think in that case the check simplifies to ensuring that the adversarial chain has at most k blocks after the intersection.

I'm considering what it would take to relax the generation and the check.

Reply:
hm, actually, if we merge #888, this is not an issue we need to fix as chains will always have @k+1@ blocks after the intersection.

I'll add a comment on this aspect in #888 instead.

Reply:
Addressed in #888 in 55f7e56 by removing the calls to `initConservative`.